### PR TITLE
Update jsdoc tests (Please provide some review about structure)

### PR DESCRIPTION
**Pull request by Ailrun**
## Description
These tests are for make JSDoc comments follow [the way they are parsed](https://github.com/jsdoc3/jsdoc/blob/b21427343c7294bbf1f14c718a390f3e955e37cb/lib/jsdoc/tag/dictionary/definitions.js#L260-L858).

Currently work in progress. Please provide your opinion freely.

### CODE REVIEW

**Comment by josteink** [comment]
Unfortunately I’ll be on vacation for about a week now with no access to a laptop.

If nobody else have looked at it by then, I’ll take a look myself.

**Comment by Ailrun** [comment]
@josteink It's almost done (I'm on alphabet 's')

**Comment by josteink** [comment]
Wow. Looking at the amount of tests added... That is crazy extensive compared to what we've had before.

Can't wait to review this thing and get it merged :)

**Comment by josteink** [comment]
Sorry about the absence. Have been moving between houses, so not much time for internets.

Is this ready to be reviewed and merged?

**Comment by Ailrun** [comment]
@josteink Could you please review about the structure first? I'm recently working on emotion typing so it will takes some time to genuinely finish this.

**Comment by josteink** [comment]
I'm happy with the overall structure of your new tests. If they were complete, I'd be happy to merge those bits right now. So all in all no major complaints.

However ... If I were to nitpick on one minor thing (and it's not even related to the structure of your tests)...

I see you've extracted the general test-helpers out into their own el-file, so that you can reuse it in a new dedicated jsdoc test-file.

That change makes the existing `typescript-mode-tests.el` a hybrid of sorts which:

1. has its own tests and
2. aggregates tests from other files.

That seems a bit inconsistent from my point of view.

IMO if we're going that path, we should extract *all* tests into dedicated test-files, and leave `typescript-mode-tests.el` a pure aggregate. We could call the new file `typescript-mode-general-tests.el` or something similar for now, if you want to avoid to categorize every single test we already have.

Does that sound reasonable?

**Comment by Ailrun** [comment]
OK, I will rename that too.

File: typescript-mode-jsdoc-tests.el [hunk 1]
```diff
@@ -0,0 +1,3086 @@
...
+     "@deprecated"
+     "@deprecated Some explanations about deprecation"
+     )
+   '(
+     ("@deprecated$" . typescript-jsdoc-tag)
+     ("@deprecated Some explanations about deprecation" . typescript-jsdoc-tag)
+     ("Some explanations about deprecation" . font-lock-comment-face)
+     ("explanations about deprecation" . font-lock-comment-face)
+     ("about deprecation" . font-lock-comment-face)
+     ("deprecation" . font-lock-comment-face)
```
Comment (Ailrun) [inline typescript-mode-jsdoc-tests.el]: Again desc-like

File: typescript-mode-jsdoc-tests.el [hunk 2]
```diff
@@ -0,0 +1,3086 @@
...
+     ;; ("inlineExample with multiple terms" . typescript-jsdoc-value)
+     ;; ("with multiple terms" . typescript-jsdoc-value)
+     ;; ("multiple terms" . typescript-jsdoc-value)
+     ;; ("terms" . typescript-jsdoc-value)
+     ("@example$" . typescript-jsdoc-tag)
+     ;; ("With multiline and multiple terms1" . typescript-jsdoc-value)
+     ;; ("multiline and multiple terms1" . typescript-jsdoc-value)
+     ;; ("and multiple terms1" . typescript-jsdoc-value)
+     ;; ("multiple terms1" . typescript-jsdoc-value)
+     ;; ("terms1" . typescript-jsdoc-value)
```
Comment (Ailrun) [inline typescript-mode-jsdoc-tests.el]: Do we have any face for code-like things?

File: typescript-mode-jsdoc-tests.el [hunk 3]
```diff
@@ -0,0 +1,3086 @@
...
+     ("@fileoverview$" . typescript-jsdoc-tag)
+     ("@fileoverview description for file1" . typescript-jsdoc-tag)
+     ("description for file1" . font-lock-comment-face)
+     ("for file1" . font-lock-comment-face)
+     ("file1" . font-lock-comment-face)
+     ("@overview$" . typescript-jsdoc-tag)
+     ("@overview description for file2" . typescript-jsdoc-tag)
+     ("description for file2" . font-lock-comment-face)
+     ("for file2" . font-lock-comment-face)
+     ("file2" . font-lock-comment-face)
```
Comment (Ailrun) [inline typescript-mode-jsdoc-tests.el]: Here goes description-like~

File: typescript-mode-jsdoc-tests.el [hunk 4]
```diff
@@ -0,0 +1,3086 @@
...
+     "@description"
+     "multi-line description about other things"
+     "@desc descInSingleWord"
+     "@desc desc about something"
+     "@desc"
+     "multi-line desc about other things"
+     )
+   '(
+     ("@description descriptionInSingleWord" . typescript-jsdoc-tag)
+     ("descriptionInSingleWord" . font-lock-comment-face)
```
[the remark this answers is not in the document]
Comment (josteink) [reply typescript-mode-jsdoc-tests.el]: IMO a comment-face is perfectly fine, unless anyone else objects.

File: typescript-mode-jsdoc-tests.el [hunk 5]
```diff
@@ -0,0 +1,3086 @@
...
+  "Test for meaningless though valid usage of @alias tags and its alias."
+  (jsdoc-multiline-test
+   '(
+     "@alias multiple words"
+     "@alias {AliasType}"
+     )
+   '(
+     ("@alias multiple words" . typescript-jsdoc-tag)
+     ("multiple words" . typescript-jsdoc-value)
+     ("words" . font-lock-comment-face)
```
[the remark this answers is not in the document]
Comment (josteink) [reply typescript-mode-jsdoc-tests.el]: I'm not sure I know what use-case you are referring to here. I would have to see a live example to have an opinion.

File: typescript-mode-jsdoc-tests.el [hunk 6]
```diff
@@ -0,0 +1,3086 @@
...
+     )
+   '(
+     ("@class$" . typescript-jsdoc-tag)
+     ("@class ExampleClass" . typescript-jsdoc-tag)
+     ;; ("ExampleClass" . typescript-jsdoc-value)
+     ("@class Class tag for description0" . typescript-jsdoc-tag)
+     ("Class tag for description0" . font-lock-comment-face)
+     ("tag for description0" . font-lock-comment-face)
+     ("for description0" . font-lock-comment-face)
+     ("description0" . font-lock-comment-face)
```
[the remark this answers is not in the document]
Comment (josteink) [reply typescript-mode-jsdoc-tests.el]: I'm 100% happy about this spec.

File: typescript-mode-jsdoc-tests.el [hunk 7]
```diff
@@ -0,0 +1,3086 @@
...
+     )
+   '(
+     ("@author Exampleauthor$" . typescript-jsdoc-tag)
+     ;; ("Exampleauthor" . typescript-jsdoc-value)
+     ("@author Exampleauthor Withfamilyname" . typescript-jsdoc-tag)
+     ;; ("Exampleauthor Withfamilyname" . typescript-jsdoc-value)
+     ;; ("Withfamilyname" . typescript-jsdoc-value)
+     ("@author Exampleauthor <example@author.email>" . typescript-jsdoc-tag)
+     ;; ("Exampleauthor <example@author.email>" . typescript-jsdoc-value)
+     ;; ("<example@author.email>" . typescript-jsdoc-value)
```
[the remark this answers is not in the document]
Comment (josteink) [reply typescript-mode-jsdoc-tests.el]: I think it's OK as a general rule that everything after the tag (unless it's a '{parameter}' gets a regular comment-face.

If you think there should be other exceptions, I'm happy to discuss those, but generally speaking, I think comment-face makes sense. At least it's what I'm accustomed to in other IDEs like Visual Studio.

File: typescript-mode-jsdoc-tests.el [hunk 8]
```diff
@@ -0,0 +1,3086 @@
...
+   '(
+     "@constructs className and extra description"
+     "@constructs {ClassType}"
+     )
+   '(
+     ("@constructs className and extra description" . typescript-jsdoc-tag)
+     ;; ("className and extra description" . typescript-jsdoc-value)
+     ("and extra description" . font-lock-comment-face)
+     ("extra description" . font-lock-comment-face)
+     ("description" . font-lock-comment-face)
```
[the remark this answers is not in the document]
Comment (josteink) [reply typescript-mode-jsdoc-tests.el]: As someone clearly more knowledgable about jsdoc than I am, what do you think would be the best behaviour?

File: typescript-mode-jsdoc-tests.el [hunk 9]
```diff
@@ -0,0 +1,3086 @@
...
+  "Test for meaningful and valid usage of @copyright tags and its alias."
+  (jsdoc-multiline-test
+   '(
+     "@copyright Some copyright texts"
+     )
+   '(
+     ("@copyright Some copyright texts" . typescript-jsdoc-tag)
+     ;; ("Some copyright texts" . typescript-jsdoc-value)
+     ;; ("copyright texts" . typescript-jsdoc-value)
+     ;; ("texts" . typescript-jsdoc-value)
```
[the remark this answers is not in the document]
Comment (josteink) [reply typescript-mode-jsdoc-tests.el]: If we treat descriptions as comments, this should be too. If we don't, then we shouldn't do that here either. We should try to be as consistent as possible.

File: typescript-mode-jsdoc-tests.el (hunk 2 above)
Comment (josteink) [reply typescript-mode-jsdoc-tests.el]: Don't think so. We could add it, if there's a use for it.

File: typescript-mode-jsdoc-tests.el (hunk 3 above)
Comment (josteink) [reply typescript-mode-jsdoc-tests.el]: Looks good to me.

**Comment by josteink** [comment]
Sorry about the late response. I'm in the process of moving between houses, and don't really have too much spare time for other projects.

Hopefully my feedback has been useful (and consistent) enough for you to make some progress.

If you feel like there's some points you still want me to address, don't hesitate to ask.

File: typescript-mode-jsdoc-tests.el (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ailrun) [reply typescript-mode-jsdoc-tests.el]: @josteink It generates hyperlink-less document when you gives multiple words.
However, it does not emit any error when you use JSDoc, and it just generates document.

File: typescript-mode-jsdoc-tests.el (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ailrun) [reply typescript-mode-jsdoc-tests.el]: If we use other face for descriptions, we can give these extra values `font-lock-comment-face`. However, it looks like it's not the case in your opinion...

[the remark this answers is not in the document]
Comment (Ailrun) [reply typescript-mode-jsdoc-tests.el]: Alright.

**Comment by Ailrun** [comment]
@josteink I think it's does for now.
I will fix codes (and uncomment tests) in another PR.

**Comment by josteink** [comment]
Sounds great. Thanks for putting in the effort!

Let me know if you have other PRs you want me to review, and I'll try to be more useful next time around :)

**Pull request merged**